hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/HazelcastSqlValidator.java

```diff
@@ -17,53 +17,284 @@
+import static org.apache.calcite.sql.type.SqlTypeName.CHAR;
+import static org.apache.calcite.sql.type.SqlTypeName.DECIMAL;
+import static org.apache.calcite.sql.type.SqlTypeName.DOUBLE;
+import static org.apache.calcite.sql.type.SqlTypeName.NULL;
+import static org.apache.calcite.sql.type.SqlTypeName.NUMERIC_TYPES;
+import static org.apache.calcite.sql.type.SqlTypeName.VARCHAR;
+import static org.apache.calcite.util.Static.RESOURCE;
+
 /**
  * Hazelcast-specific SQL validator.
  */
 public class HazelcastSqlValidator extends SqlValidatorImpl {
-    public HazelcastSqlValidator(
-        SqlOperatorTable opTab,
-        SqlValidatorCatalogReader catalogReader,
-        RelDataTypeFactory typeFactory,
-        SqlConformance conformance
-    ) {
-        super(opTab, catalogReader, typeFactory, Config.DEFAULT.withSqlConformance(conformance));
+
+    private static final Config CONFIG = Config.DEFAULT.withIdentifierExpansion(true);
+
+    private final Map<SqlNode, RelDataType> knownNodeTypes = new IdentityHashMap<>();
+
+    public HazelcastSqlValidator(SqlOperatorTable opTab, SqlValidatorCatalogReader catalogReader,
+                                 HazelcastTypeFactory typeFactory, SqlConformance conformance) {
+        super(opTab, catalogReader, typeFactory, CONFIG.withSqlConformance(conformance));
+        setTypeCoercion(new HazelcastTypeCoercion(this));
+    }
+
+    /**
+     * Sets {@code type} as the known type for {@code node}.
+     *
+     * @param node the node to set the known type of.
+     * @param type the type to set the know node type to.
+     */
+    public void setKnownNodeType(SqlNode node, RelDataType type) {
+        assert !getUnknownType().equals(type);
+        knownNodeTypes.put(node, type);
+    }
+
+    /**
+     * Obtains a type known by this validator for the given node.
+     *
+     * @param node the node to obtain the type of.
+     * @return the node type known by this validator or {@code null} if the type
+     * of the given node is not known yet.
+     */
+    public RelDataType getKnownNodeType(SqlNode node) {
+        return knownNodeTypes.get(node);
+    }
+
+    @Override
+    public void validateQuery(SqlNode node, SqlValidatorScope scope, RelDataType targetRowType) {
+        super.validateQuery(node, scope, targetRowType);
+
+        if (node instanceof SqlSelect) {
+            // Derive the types for offset-fetch expressions, Calcite doesn't do
+            // that automatically.
+
+            SqlSelect select = (SqlSelect) node;
+
+            SqlNode offset = select.getOffset();
+            if (offset != null) {
+                deriveType(scope, offset);
+            }
+
+            SqlNode fetch = select.getFetch();
+            if (fetch != null) {
+                deriveType(scope, fetch);
+            }
+        }
+    }
+
+    @Override
+    public SqlNode validate(SqlNode topNode) {
+        if (topNode.getKind().belongsTo(SqlKind.DDL)) {
```

Comment:
DDL is not supported in this PR, so the whole `validate` method could be removed safely.

Reply:
fixrd in 417d1eb990bc8c0d842ea8172b8bb27f529784b0